btcutil/tx.go

```diff
@@ -37,30 +38,69 @@ func (t *Tx) MsgTx() *wire.MsgTx {
 
 // Hash returns the hash of the transaction.  This is equivalent to
 // calling TxHash on the underlying wire.MsgTx, however it caches the
-// result so subsequent calls are more efficient.
+// result so subsequent calls are more efficient.  If the Tx has the
+// raw bytes of the tx cached, it will use that and skip serialization.
 func (t *Tx) Hash() *chainhash.Hash {
 	// Return the cached hash if it has already been generated.
 	if t.txHash != nil {
 		return t.txHash
 	}
 
 	// Cache the hash and return it.
-	hash := t.msgTx.TxHash()
+	var hash chainhash.Hash
+
+	// If we have the raw bytes, then don't call msgTx.TxHash as that has the
+	// overhead of serialization.
+	if len(t.rawBytes) > 0 {
+		// If the raw bytes contain the witness, we must strip it out before
+		// calculating the hash.
+		if t.HasWitness() {
+			baseSize := t.msgTx.SerializeSizeStripped()
+			bytes := make([]byte, 0, baseSize)
```

Comment:
nit: `result` or `nonWitnessBytes` to not collide with the imported package name?